File: eporner.js
const axios = require('axios');
const cheerio = require('cheerio');

module.exports = async function eporner(query) {
  const url = `https://www.eporner.com/search/${encodeURIComponent(query)}/`;
  const results = [];

  try {
    const { data } = await axios.get(url);
    const $ = cheerio.load(data);

    // Try multiple selectors for video cards
    let videoEls = $('.mb-3, .video-thumb, .thumb, .video-block, .video-item');
    if (videoEls.length === 0) {
      videoEls = $('a[href*="/video/"]'); // fallback: links to videos
    }

    videoEls.each((i, el) => {
      let title = $(el).find('img').attr('alt')?.trim() || $(el).find('.title').text().trim() || $(el).attr('title') || $(el).find('a').attr('title') || $(el).text().trim();
      let href = $(el).find('a').attr('href') || $(el).attr('href');
      let duration = $(el).find('.text-white').text().trim() || $(el).find('.duration').text().trim();
      let preview = $(el).find('img').attr('data-src') || $(el).find('img').attr('src') || '';

      // Normalize href
      if (href && !href.startsWith('http')) {
        href = 'https://www.eporner.com' + href;
      }
      // Normalize preview
      if (preview && preview.startsWith('//')) preview = 'https:' + preview;
      if (preview && preview.startsWith('/')) preview = 'https://www.eporner.com' + preview;

      if (title && href) {
        results.push({
          title,
          url: href,
          duration,
          preview,
          source: "Eporner"
        });
      }
    });

    if (results.length === 0) {
      console.error("eporner: No results found. Selectors may be outdated or site structure changed.");
    }
  } catch (err) {
    console.error("eporner error:", err.message);
  }

  return results;
};

File: scrapers/eporner.py
import requests
from bs4 import BeautifulSoup

def scrape_eporner(query, mode="straight", page=1):
    results = []
    seen = set()
    # Use ScraperAPI to bypass anti-bot
    max_pages = 50
    max_results = 1000
    for p in range(page, page + max_pages):
        # No gay subdomain, but you can bias keywords
        if mode == "gay":
            url = f"https://www.eporner.com/search/{p}/?q=gay+{query}"
        elif mode == "trans":
            url = f"https://www.eporner.com/search/{p}/?q=trans+{query}"
        else:
            url = f"https://www.eporner.com/search/{p}/?q={query}"
        try:
            payload = {
                'api_key': '23c1327aeb270f44bb141d469c7f9823',
                'url': url,
                'country_code': 'us',
            }
            r = requests.get('https://api.scraperapi.com/', params=payload, timeout=20)
            if r.status_code != 200:
                print(f"[Eporner] Non-200 status: {r.status_code} for {url}")
                continue
            soup = BeautifulSoup(r.content, "html.parser")
            found = False
            for vid in soup.select("div.mb-video, div.mb-3, .video-thumb"):
                try:
                    a = vid.select_one("a")
                    if not a or not a.has_attr('href'):
                        continue
                    title = a.get('title') or a.text.strip()
                    href = a.get('href')
                    if isinstance(href, list):
                        href = href[0] if href else None
                    if not href:
                        continue
                    video_url = f"https://www.eporner.com{href}"
                    img = vid.select_one("img")
                    preview = img.get('data-src') or img.get('src') if img else ""
                    if isinstance(preview, list):
                        preview = preview[0] if preview else ""
                    preview = str(preview)
                    if preview.startswith('//'):
                        preview = 'https:' + preview
                    if not preview or "placeholder" in preview:
                        continue
                    # Deduplicate by video URL
                    if video_url in seen:
                        continue
                    seen.add(video_url)
                    results.append({
                        "title": title,
                        "url": video_url,
                        "preview": preview,
                        "source": f"Eporner ({mode})"
                    })
                    found = True
                    if len(results) >= max_results:
                        return results
                except Exception as e:
                    print(f"[Eporner] Error parsing video: {e}")
                    continue
            if len(results) >= max_results:
                return results
            if not found:
                print(f"[Eporner] No results found on page {p}. Selector may be outdated or site structure changed.")
        except Exception as e:
            print(f"[Eporner] Error on page {p}: {e}")
    return results
